File: app/src/test/java/com/coinninja/coinkeeper/service/runner/InviteContactRunnerTest.java
package com.coinninja.coinkeeper.service.runner;

import com.coinninja.coinkeeper.TestCoinKeeperApplication;
import com.coinninja.coinkeeper.cn.wallet.CNWalletManager;
import com.coinninja.coinkeeper.model.PhoneNumber;
import com.coinninja.coinkeeper.model.dto.PendingInviteDTO;
import com.coinninja.coinkeeper.model.helpers.InviteTransactionSummaryHelper;
import com.coinninja.coinkeeper.service.client.SignedCoinKeeperApiClient;
import com.coinninja.coinkeeper.service.client.model.Contact;
import com.coinninja.coinkeeper.service.client.model.InvitedContact;
import com.coinninja.coinkeeper.util.Intents;
import com.coinninja.coinkeeper.util.PhoneNumberUtil;
import com.coinninja.coinkeeper.util.currency.USDCurrency;
import com.coinninja.coinkeeper.util.uuid.UUIDGenerator;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import okhttp3.MediaType;
import okhttp3.ResponseBody;
import retrofit2.Response;

import static org.mockito.Matchers.any;
import static org.mockito.Matchers.anyLong;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(RobolectricTestRunner.class)
@Config(application = TestCoinKeeperApplication.class)
public class InviteContactRunnerTest {

    @Mock
    private SignedCoinKeeperApiClient client;
    @Mock
    private InviteContactRunner.OnInviteListener onInviteListener;
    @Mock
    CNWalletManager cnWalletManager;
    @Mock
    private UUIDGenerator generator;

    @Mock
    PendingInviteDTO pendingInviteDTO;

    @Mock
    InviteTransactionSummaryHelper inviteTransactionSummaryHelper;

    private long bitcoinUSDPrice;
    private long satoshisSending;
    private String sendBtcToPhoneNumber;
    private InviteContactRunner inviteContactRunner;
    private Contact myContact;
    private final PhoneNumberUtil phoneNumberUtil = new PhoneNumberUtil();
    private PhoneNumber toPhoneNumber;
    private PhoneNumber accountPhoneNumber;
    private String uuid = "23742734-23742734-23472734-23742734";

    @Before
    public void setUp() throws Exception {
        MockitoAnnotations.initMocks(this);
        String myAccountPhoneNumber = "2223334444";
        sendBtcToPhoneNumber = "2165454564";
        toPhoneNumber = new PhoneNumber(sendBtcToPhoneNumber);
        accountPhoneNumber = new PhoneNumber(myAccountPhoneNumber);
        satoshisSending = 4568949465L;
        bitcoinUSDPrice = 800000;

        pendingInviteDTO = mock(PendingInviteDTO.class);
        when(pendingInviteDTO.getInviteAmount()).thenReturn(satoshisSending);
        when(pendingInviteDTO.getBitcoinPrice()).thenReturn(bitcoinUSDPrice);
        when(pendingInviteDTO.getRequestId()).thenReturn(uuid);

        myContact = new Contact(accountPhoneNumber, "send_to_self", true);
        when(cnWalletManager.getContact()).thenReturn(myContact);
        when(generator.generate()).thenReturn(uuid);
        inviteContactRunner = new InviteContactRunner(client, cnWalletManager, phoneNumberUtil, inviteTransactionSummaryHelper);
        inviteContactRunner.setPendingInviteDTO(pendingInviteDTO);
    }

    @After
    public void tearDown() throws Exception {
        client = null;
        onInviteListener = null;
        cnWalletManager = null;
        sendBtcToPhoneNumber = null;
        inviteContactRunner = null;
        myContact = null;
        pendingInviteDTO = null;
    }

    @Test
    public void send_invite_to_contact_test() {
        Contact sendToPhoneNumber = new Contact(toPhoneNumber, "Bill Bob", false);
        InvitedContact contact = mock(InvitedContact.class);

        Response mockResponse = Response.success(contact);

        when(client.invitePhoneNumber(anyLong(), anyLong(), any(), any(), any())).thenReturn(mockResponse);
        inviteContactRunner.setOnInviteListener(onInviteListener);

        inviteContactRunner.execute(sendToPhoneNumber);

        PhoneNumber senderPhoneNumber = myContact.getPhoneNumber();
        PhoneNumber receiverPhoneNumber = sendToPhoneNumber.getPhoneNumber();
        verify(client).invitePhoneNumber(36551596l, 4568949465l,
                senderPhoneNumber, receiverPhoneNumber, uuid);
        verify(onInviteListener).onInviteSuccessful(contact);
    }

    @Test
    public void send_invite_to_contact_fail() {
        Contact sendToPhoneNumber = new Contact(toPhoneNumber, "Bill Bob", false);

        ResponseBody body = ResponseBody.create(MediaType.parse("text"), "bad request");
        Response mockResponse = Response.error(400, body);

        when(client.invitePhoneNumber(anyLong(), anyLong(), any(), any(), any())).thenReturn(mockResponse);
        inviteContactRunner.setOnInviteListener(onInviteListener);

        inviteContactRunner.execute(sendToPhoneNumber);

        PhoneNumber senderPhoneNumber = myContact.getPhoneNumber();
        PhoneNumber receiverPhoneNumber = sendToPhoneNumber.getPhoneNumber();
        verify(client).invitePhoneNumber(36551596l, 4568949465l,
                senderPhoneNumber, receiverPhoneNumber, uuid);

        verify(onInviteListener).onInviteError(Intents.ACTION_DROPBIT__ERROR_UNKNOWN, "bad request");
    }

    @Test
    public void send_dropbit_rate_limit_error() {
        Contact sendToPhoneNumber = new Contact(toPhoneNumber, "Bill Bob", false);

        ResponseBody body = ResponseBody.create(MediaType.parse("text"), "rate limit error");
        Response mockResponse = Response.error(429, body);

        when(client.invitePhoneNumber(anyLong(), anyLong(), any(), any(), any())).thenReturn(mockResponse);
        inviteContactRunner.setOnInviteListener(onInviteListener);

        inviteContactRunner.execute(sendToPhoneNumber);

        PhoneNumber senderPhoneNumber = myContact.getPhoneNumber();
        PhoneNumber receiverPhoneNumber = sendToPhoneNumber.getPhoneNumber();
        verify(client).invitePhoneNumber(36551596l, 4568949465l,
                senderPhoneNumber, receiverPhoneNumber, uuid);

        verify(onInviteListener).onInviteError(Intents.ACTION_DROPBIT__ERROR_RATE_LIMIT, "rate limit error");
    }
}

File: app/src/test/java/com/coinninja/coinkeeper/service/runner/InviteContactRunnerTest.kt
package com.coinninja.coinkeeper.service.runner

import androidx.test.ext.junit.runners.AndroidJUnit4
import com.coinninja.coinkeeper.model.Identity
import com.coinninja.coinkeeper.model.db.DropbitMeIdentity
import com.coinninja.coinkeeper.model.db.enums.IdentityType
import com.coinninja.coinkeeper.model.dto.PendingInviteDTO
import com.coinninja.coinkeeper.service.client.SignedCoinKeeperApiClient
import com.coinninja.coinkeeper.service.client.model.InviteUserPayload
import com.coinninja.coinkeeper.service.client.model.InvitedContact
import com.coinninja.coinkeeper.util.DropbitIntents
import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.argumentCaptor
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever
import okhttp3.MediaType
import okhttp3.ResponseBody
import org.hamcrest.CoreMatchers.equalTo
import org.hamcrest.MatcherAssert.assertThat
import org.junit.Assert.assertNull
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mockito.verify
import retrofit2.Response

@RunWith(AndroidJUnit4::class)
class InviteContactRunnerTest {

    private fun createRunner(): InviteContactRunner {
        val inviteContactRunner = InviteContactRunner(mock(), mock(), mock(), mock())

        val uuid = "23742734-23742734-23472734-23742734"
        val identity = Identity(identityType = IdentityType.PHONE, hash = "--hash--", value = "+13305551111")
        val pendingInviteDTO = PendingInviteDTO(identity, 50000L, 10000L, 100L, "", false, uuid)
        val phoneIdentity: DropbitMeIdentity = mock()
        whenever(phoneIdentity.type).thenReturn(IdentityType.PHONE)
        whenever(phoneIdentity.identity).thenReturn("+13305551111")
        whenever(inviteContactRunner.dropbitAccountHelper.phoneIdentity()).thenReturn(phoneIdentity)
        inviteContactRunner.setPendingInviteDTO(pendingInviteDTO)
        return inviteContactRunner
    }

    @Test
    fun send_invite_to_contact_test() {
        val runner = createRunner()
        val contact: InvitedContact = mock()
        val identity = Identity(IdentityType.PHONE, "+13305551111", "--hash--", "Some User")
        val mockResponse = Response.success(contact)
        whenever(runner.client.inviteUser(any())).thenReturn(mockResponse)
        runner.setOnInviteListener(mock())

        runner.execute(identity)

        verify(runner.onInviteListener!!).onInviteSuccessful(contact)
    }

    @Test
    fun send_invite_to_contact_fail() {
        val identity = Identity(IdentityType.PHONE, "+13305551111", "--hash--", "Some User")
        val runner = createRunner()

        val body = ResponseBody.create(MediaType.parse("text"), "bad request")
        val mockResponse = Response.error<InvitedContact>(400, body)

        whenever(runner.client.inviteUser(any())).thenReturn(mockResponse)
        runner.setOnInviteListener(mock())

        runner.execute(identity)

        verify(runner.onInviteListener!!).onInviteError(DropbitIntents.ACTION_DROPBIT__ERROR_UNKNOWN, "bad request")
    }

    @Test
    fun send_dropbit_rate_limit_error() {
        val identity = Identity(IdentityType.PHONE, "+13305551111", "--hash--", "Some User")
        val runner = createRunner()
        val body = ResponseBody.create(MediaType.parse("text"), "rate limit error")
        val mockResponse = Response.error<InvitedContact>(429, body)
        whenever(runner.client.inviteUser(any())).thenReturn(mockResponse)
        runner.setOnInviteListener(mock())

        runner.execute(identity)

        verify(runner.onInviteListener!!).onInviteError(DropbitIntents.ACTION_DROPBIT__ERROR_RATE_LIMIT, "rate limit error")
    }

    @Test
    fun invites_twitter_with_handle_from_twitter() {
        val runner = createRunner()
        whenever(runner.client.inviteUser(any())).thenReturn(Response.success(mock<InvitedContact>()))
        val captor = argumentCaptor<InviteUserPayload>()
        whenever(runner.dropbitAccountHelper.isTwitterVerified).thenReturn(true)
        val twitterIdentity: DropbitMeIdentity = mock()
        whenever(runner.dropbitAccountHelper.twitterIdentity()).thenReturn(twitterIdentity)
        whenever(twitterIdentity.handle).thenReturn("MyHandle")
        whenever(twitterIdentity.identity).thenReturn("5432112345")
        whenever(twitterIdentity.type).thenReturn(IdentityType.TWITTER)
        val identity = Identity(IdentityType.TWITTER, "1234567890", "1234567890", "Some User")
        runner.execute(identity)

        verify(runner.client).inviteUser(captor.capture())
        val invitedUser = captor.firstValue

        assertThat(invitedUser.amount.btc, equalTo(10000L))
        assertThat(invitedUser.amount.usd, equalTo(5L))

        assertThat(invitedUser.receiver.handle, equalTo("Some User"))
        assertThat(invitedUser.receiver.identity, equalTo("1234567890"))
        assertThat(invitedUser.receiver.type, equalTo("twitter"))

        assertThat(invitedUser.sender.handle, equalTo("MyHandle"))
        assertThat(invitedUser.sender.identity, equalTo("5432112345"))
        assertThat(invitedUser.sender.type, equalTo("twitter"))
    }

    @Test
    fun invites_phone_with_out_handle_from_phone() {
        val runner = createRunner()
        whenever(runner.client.inviteUser(any())).thenReturn(Response.success(mock()))
        val captor = argumentCaptor<InviteUserPayload>()
        whenever(runner.dropbitAccountHelper.isPhoneVerified).thenReturn(true)
        val phoneIdentity: DropbitMeIdentity = mock()
        whenever(runner.dropbitAccountHelper.phoneIdentity()).thenReturn(phoneIdentity)
        whenever(phoneIdentity.identity).thenReturn("+13305551111")
        whenever(phoneIdentity.type).thenReturn(IdentityType.PHONE)
        val identity = Identity(IdentityType.PHONE, "+13305550000", "--hash--", null)
        runner.execute(identity)

        verify(runner.client).inviteUser(captor.capture())
        val (amount, sender, receiver) = captor.firstValue

        assertThat(amount.btc, equalTo(10000L))
        assertThat(amount.usd, equalTo(5L))

        assertNull(receiver.handle)
        assertThat(receiver.identity, equalTo("13305550000"))
        assertThat(receiver.type, equalTo("phone"))

        assertNull(sender.handle)
        assertThat(sender.identity, equalTo("13305551111"))
        assertThat(sender.type, equalTo("phone"))
    }

    @Test
    fun `invites phone from twitter when phone is not verified`() {
        val runner = createRunner()
        whenever(runner.client.inviteUser(any())).thenReturn(Response.success(mock()))
        val captor = argumentCaptor<InviteUserPayload>()
        val twitterIdentity: DropbitMeIdentity = mock()
        whenever(runner.dropbitAccountHelper.isPhoneVerified).thenReturn(false)
        whenever(runner.dropbitAccountHelper.twitterIdentity()).thenReturn(twitterIdentity)
        whenever(twitterIdentity.handle).thenReturn("MyHandle")
        whenever(twitterIdentity.identity).thenReturn("5432112345")
        whenever(twitterIdentity.type).thenReturn(IdentityType.TWITTER)
        val identity = Identity(IdentityType.PHONE, "+13305550000", "--hash--", null)
        runner.execute(identity)

        verify<SignedCoinKeeperApiClient>(runner.client).inviteUser(captor.capture())
        val (amount, sender, receiver) = captor.firstValue

        assertThat(amount.btc, equalTo(10000L))
        assertThat(amount.usd, equalTo(5L))

        assertNull(receiver.handle)
        assertThat(receiver.identity, equalTo("13305550000"))
        assertThat(receiver.type, equalTo("phone"))

        assertThat(sender.handle, equalTo("MyHandle"))
        assertThat(sender.identity, equalTo("5432112345"))
        assertThat(sender.type, equalTo("twitter"))
    }
}